translate-c: handle string concatenation of function calls

### Zig Version

0.12.0-dev.3366+8e7d9afda

### Steps to Reproduce and Observed Behavior

```bash
# See https://github.com/willnode/summon/tree/a246528
git clone https://github.com/willnode/summon.git -b try-zig
cd summon
make install-nginx
zig build
```

```
install
└─ install summon
   └─ zig build-lib summon Debug native 3 errors
/home/willnode/summon/zig-cache/o/9f471bc5f7934470be5987871f47d714/cimport.zig:4488:11: error: opaque types have unknown size and therefore cannot be directly embedded in structs
    file: ngx_file_t = @import("std").mem.zeroes(ngx_file_t),
          ^~~~~~~~~~
/home/willnode/summon/zig-cache/o/9f471bc5f7934470be5987871f47d714/cimport.zig:3047:31: note: opaque declared here
pub const struct_ngx_file_s = opaque {};
                              ^~~~~~~~~
/home/willnode/summon/zig-cache/o/9f471bc5f7934470be5987871f47d714/cimport.zig:5842:36: error: unable to translate C expr: unexpected token 'a string literal'
pub const NGX_MODULE_SIGNATURE_0 = @compileError("unable to translate C expr: unexpected token 'a string literal'"); // /home/willnode/summon/nginx-1.22.1/src/core/ngx_module.h:21:9
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.zig:24:13: error: non-extern variable with opaque type 'cimport.struct_ngx_buf_s'
    var b: c.ngx_buf_t = undefined;
           ~^~~~~~~~~~
/home/willnode/summon/zig-cache/o/9f471bc5f7934470be5987871f47d714/cimport.zig:3049:30: note: opaque declared here
pub const struct_ngx_buf_s = opaque {};
                             ^~~~~~~~~
error: the following command failed with 3 compilation errors:
/home/willnode/.local/opt/zig-v0.12.0-dev.3366+8e7d9afda/zig build-lib -ODebug -I /home/willnode/summon/nginx-1.22.1/objs -I /home/willnode/summon/nginx-1.22.1/src/mail -I /home/willnode/summon/nginx-1.22.1/src/os/unix -I /home/willnode/summon/nginx-1.22.1/src/os -I /home/willnode/summon/nginx-1.22.1/src/event/modules -I /home/willnode/summon/nginx-1.22.1/src/event -I /home/willnode/summon/nginx-1.22.1/src/stream -I /home/willnode/summon/nginx-1.22.1/src/http/v2 -I /home/willnode/summon/nginx-1.22.1/src/http/modules/perl -I /home/willnode/summon/nginx-1.22.1/src/http/modules -I /home/willnode/summon/nginx-1.22.1/src/http -I /home/willnode/summon/nginx-1.22.1/src/core -I /home/willnode/summon/nginx-1.22.1/src/misc -I /home/willnode/summon/nginx-1.22.1/objs/src/mail -I /home/willnode/summon/nginx-1.22.1/objs/src/os/unix -I /home/willnode/summon/nginx-1.22.1/objs/src/os -I /home/willnode/summon/nginx-1.22.1/objs/src/event/modules -I /home/willnode/summon/nginx-1.22.1/objs/src/event -I /home/willnode/summon/nginx-1.22.1/objs/src/stream -I /home/willnode/summon/nginx-1.22.1/objs/src/http/v2 -I /home/willnode/summon/nginx-1.22.1/objs/src/http/modules/perl -I /home/willnode/summon/nginx-1.22.1/objs/src/http/modules -I /home/willnode/summon/nginx-1.22.1/objs/src/http -I /home/willnode/summon/nginx-1.22.1/objs/src/core -I /home/willnode/summon/nginx-1.22.1/objs/src/misc -Mroot=/home/willnode/summon/src/main.zig -lc --cache-dir /home/willnode/summon/zig-cache --global-cache-dir /home/willnode/.cache/zig --name summon -dynamic --listen=- 
Build Summary: 0/3 steps succeeded; 1 failed (disable with --summary none)
install transitive failure
└─ install summon transitive failure
   └─ zig build-lib summon Debug native 3 errors
error: the following build command failed with exit code 1:
/home/willnode/summon/zig-cache/o/a9ce6dcf4254ec6f5921d0f7e1cffe37/build /home/willnode/.local/opt/zig-v0.12.0-dev.3366+8e7d9afda/zig /home/willnode/summon /home/willnode/summon/zig-cache /home/willnode/.cache/zig --seed 0x1b9ba675 -Z53dd1933cc77c915
```

### Expected Behavior

Library should be successfully compiled

--- a/src/translate_c.zig
+++ b/src/translate_c.zig
@@ -5724,7 +5724,7 @@ fn escapeUnprintables(ctx: *Context, m: *MacroCtx) ![]const u8 {
     };
 }
 
-fn parseCPrimaryExprInner(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!Node {
+fn parseCPrimaryExpr(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!Node {
     const tok = m.next().?;
     const slice = m.slice();
     switch (tok) {
@@ -5754,7 +5754,7 @@ fn parseCPrimaryExprInner(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!N
         },
         .identifier, .extended_identifier => {
             if (c.global_scope.blank_macros.contains(slice)) {
-                return parseCPrimaryExprInner(c, m, scope);
+                return parseCPrimaryExpr(c, m, scope);
             }
             const mangled_name = scope.getAlias(slice);
             if (builtin_typedef_map.get(mangled_name)) |ty| return Tag.type.create(c.arena, ty);
@@ -5779,35 +5779,6 @@ fn parseCPrimaryExprInner(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!N
             return error.ParseError;
         },
     }
-}
-
-fn parseCPrimaryExpr(c: *Context, m: *MacroCtx, scope: *Scope) ParseError!Node {
-    var node = try parseCPrimaryExprInner(c, m, scope);
-    // In C the preprocessor would handle concatting strings while expanding macros.
-    // This should do approximately the same by concatting any strings and identifiers
-    // after a primary expression.
-    while (true) {
-        switch (m.peek().?) {
-            .string_literal,
-            .string_literal_utf_16,
-            .string_literal_utf_8,
-            .string_literal_utf_32,
-            .string_literal_wide,
-            => {},
-            .identifier, .extended_identifier => {
-                const tok = m.list[m.i + 1];
-                const slice = m.source[tok.start..tok.end];
-                if (c.global_scope.blank_macros.contains(slice)) {
-                    m.i += 1;
-                    continue;
-                }
-            },
-            else => break,
-        }
-        const rhs = try parseCPrimaryExprInner(c, m, scope);
-        node = try Tag.array_cat.create(c.arena, .{ .lhs = node, .rhs = rhs });
-    }
-    return node;
 }
 
 fn macroIntFromBool(c: *Context, node: Node) !Node {
@@ -6241,6 +6212,35 @@ fn parseCAbstractDeclarator(c: *Context, m: *MacroCtx, node: Node) ParseError!No
 }
 
 fn parseCPostfixExpr(c: *Context, m: *MacroCtx, scope: *Scope, type_name: ?Node) ParseError!Node {
+    var node = try parseCPostfixExprInner(c, m, scope, type_name);
+    // In C the preprocessor would handle concatting strings while expanding macros.
+    // This should do approximately the same by concatting any strings and identifiers
+    // after a primary or postfix expression.
+    while (true) {
+        switch (m.peek().?) {
+            .string_literal,
+            .string_literal_utf_16,
+            .string_literal_utf_8,
+            .string_literal_utf_32,
+            .string_literal_wide,
+            => {},
+            .identifier, .extended_identifier => {
+                const tok = m.list[m.i + 1];
+                const slice = m.source[tok.start..tok.end];
+                if (c.global_scope.blank_macros.contains(slice)) {
+                    m.i += 1;
+                    continue;
+                }
+            },
+            else => break,
+        }
+        const rhs = try parseCPostfixExprInner(c, m, scope, type_name);
+        node = try Tag.array_cat.create(c.arena, .{ .lhs = node, .rhs = rhs });
+    }
+    return node;
+}
+
+fn parseCPostfixExprInner(c: *Context, m: *MacroCtx, scope: *Scope, type_name: ?Node) ParseError!Node {
     var node = type_name orelse try parseCPrimaryExpr(c, m, scope);
     while (true) {
         switch (m.next().?) {

--- a/test/cases/translate_c/macro_function_string_concat.c
+++ b/test/cases/translate_c/macro_function_string_concat.c
@@ -1,0 +1,11 @@
+#define bar() ""
+#define FOO bar() "," bar()
+
+// translate-c
+// target=x86_64-linux
+// c_frontend=clang
+//
+// pub inline fn bar() @TypeOf("") {
+//     return "";
+// }
+// pub const FOO = bar() ++ "," ++ bar();

